astropy/units/tests/test_quantity_annotations.py

```diff
@@ -1,9 +1,112 @@
 # -*- coding: utf-8 -*-
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
+# STDLIB
+import typing as T
+
+# THIRD PARTY
 import pytest
 
+# LOCAL
 from astropy import units as u  # pylint: disable=W0611
+from astropy.units import Quantity
+from astropy.units._typing import HAS_ANNOTATED, Annotated
+
+
+def test_ignore_generic_type_annotations():
+    """Test annotations that are not unit related are ignored.
+
+    This test passes if the function works.
+    """
+    # one unit, one not (should be ignored)
+    @u.quantity_input
+    def func(x: u.m, y: str):
+        return x, y
+
+    i_q, i_str = 2 * u.m, "cool string"
+    o_q, o_str = func(i_q, i_str)  # if this doesn't fail, it worked.
+    assert i_q == o_q
+    assert i_str == o_str
+
+
+@pytest.mark.skipif(not HAS_ANNOTATED, reason="need `Annotated`")
+class TestQuantityUnitAnnotations:
+    """Test Quantity[Unit] type annotation."""
+
+    def test_simple_annotation(self):
+        @u.quantity_input
+        def func(x: Quantity[u.m], y: str):
+            return x, y
+
+        i_q, i_str = 2 * u.m, "cool string"
+        o_q, o_str = func(i_q, i_str)
+        assert i_q == o_q
+        assert i_str == o_str
+
+    def test_multiple_annotation(self):
+        @u.quantity_input
+        def multi_func(a: Quantity[u.km]) -> Quantity[u.m]:
+            return a.to(u.m)
```

Comment:
Should this just return `a`? If I understood the code right, the return annotation will actually enforce the unit, so the test would still pass.

Reply:
About this... I personally prefer if the code just validated the value, not enforced units, but I didn't want to change the behavior.
Part of the reason I prefer just validation is consistency with the behavior for the inputs. Also the ``quantity_input`` refactor is going to enable unit enforcement on inputs and outputs in a subclass decorator. Also fine-grained user control on a per-input / output basis.

Thoughts?

Reply:
Agreed that `-> Quantity['length']` should be possible too, and that `quantity_input` should allow more control:
1. Check quantities of right physical type
2. Convert input/output to specific units
3. As 2, but drop to array, and attach unit on output (#7368)

For now, fine to leave the test as is!

Reply:
SGTM. Something to consider for #10655